### PR TITLE
CDAP-4454 setting output schema for sources

### PR DESCRIPTION
**Pull request by gsps1**
set output schema for sources. 
https://issues.cask.co/browse/CDAP-4454

### CODE REVIEW

File: cassandra-plugins/src/main/java/co/cask/hydrator/plugin/batch/source/BatchCassandraSource.java [hunk 1]
```diff
@@ -75,6 +76,17 @@ public BatchCassandraSource(CassandraSourceConfig config) {
   }
 
   @Override
+  public void configurePipeline(PipelineConfigurer pipelineConfigurer) {
+    Preconditions.checkArgument(!Strings.isNullOrEmpty(config.schema), "Schema must be specified.");
+    try {
+      Schema schema = Schema.parseJson(config.schema);
+      pipelineConfigurer.getStageConfigurer().setOutputSchema(schema);
+    } catch (Exception e) {
+      throw new IllegalArgumentException("Invalid input schema: " + e.getMessage(), e);
```
Comment (albertshau) [inline cassandra-plugins/src/main/java/co/cask/hydrator/plugin/batch/source/BatchCassandraSource.java]: Invalid output schema

Comment (albertshau) [inline cassandra-plugins/src/main/java/co/cask/hydrator/plugin/batch/source/BatchCassandraSource.java]: similar comment for others

File: core-plugins/src/main/java/co/cask/hydrator/plugin/batch/source/StreamBatchSource.java [hunk 2]
```diff
@@ -94,6 +94,10 @@
   public void configurePipeline(PipelineConfigurer pipelineConfigurer) {
     streamBatchConfig.validate();
     pipelineConfigurer.addStream(new Stream(streamBatchConfig.name));
+    // if no format is specified then default schema is used, if otherwise its based on input records.
+    if (streamBatchConfig.format == null) {
+      pipelineConfigurer.getStageConfigurer().setOutputSchema(DEFAULT_SCHEMA);
+    }
```
Comment (albertshau) [inline core-plugins/src/main/java/co/cask/hydrator/plugin/batch/source/StreamBatchSource.java:107]: many formats will allow or require specifying a schema, which we can use to set the outputschema.

Can use `streamBatchConfig.getFormatSpec()`, then if the spec's format is non-null, set the output schema accordingly.  One thing that is a little tricky is that the spec's schema is the schema of the body, so you'd have to add ts and headers as fields before setting the output schema.

File: mongodb-plugins/src/main/java/co/cask/hydrator/plugin/batch/source/MongoDBBatchSource.java [hunk 3]
```diff
@@ -64,6 +64,7 @@ public void configurePipeline(PipelineConfigurer pipelineConfigurer) {
     super.configurePipeline(pipelineConfigurer);
     try {
       BSONConverter.validateSchema(Schema.parseJson(config.schema));
+      pipelineConfigurer.getStageConfigurer().setOutputSchema(Schema.parseJson(config.schema));
     } catch (IOException e) {
       throw new IllegalArgumentException("Invalid schema", e);
```
Comment (albertshau) [inline mongodb-plugins/src/main/java/co/cask/hydrator/plugin/batch/source/MongoDBBatchSource.java]: lets include e.getMessage() to be consistent with other plugins.

**Comment by albertshau** [comment]
mostly the one comment about stream source. Other comments are minor.

File: core-plugins/src/main/java/co/cask/hydrator/plugin/batch/source/StreamBatchSource.java (hunk 2 above)
Comment (gsps1) [inline core-plugins/src/main/java/co/cask/hydrator/plugin/batch/source/StreamBatchSource.java:107]: yeah, have fixed it, the schema fields are added as separate fields and not as a single body field in transform, so doing the same.

**Comment by gsps1** [comment]
have addressed the comments, thanks

**Comment by albertshau** [comment]
:+1:

**Pull request merged**